### DISCUSSION TITLE
Use either Email or Username to Login

### DISCUSSION THREAD

**Discussion by EliMariaaaa** [codeigniter4/shield#1206] [Q&A]
Hello!

Can I use either email or username to log in? Currently, it's either of the two, but not both. Can I make this work without adding a select option if the user wants to log in using email or username?
<img width="386" alt="Screenshot 2024-10-13 at 3 36 19 AM" src="https://github.com/user-attachments/assets/57e4f5df-e4fb-4ab5-bdf8-deec3bafbafb">

**Comment by warcooft** [comment]
At the moment, there is no way to achieve this directly. If possible, you may need to customize Auth config and extend the `LoginAction` controller.

**Comment by grimpirate** [comment]
Here's how I go about it assuming you have the default setup to login via email:

1. Auth.validFields should include both 'username' and 'email'
2. Create a custom view under App\Views\auth\login.php which basically copies the default login form but changes the type="email" attribute to type="text"
3. Auth.views.login should be set to \App\Views\auth\login.php
4. Create a custom controller under App\Controllers\Auth\LoginController.php which extends CodeIgniter\Shield\Controllers\LoginController and overrides its loginAction method (copy the method directly from source)
5. Modify loginAction's code to unset the 'valid_email' requirement from the email rules
6. In the credentials section check the email for the presence of the '@' symbol, if it's included do nothing, otherwise set `$credentials['username'] = $credentials['email'];` and `unset($credentials['email']);`
7. Finally, configure your App\Config\Routes.php with the following:
```
service('auth')->routes($routes, ['except' => ['login']]);
$routes->get('login', '\CodeIgniter\Shield\Controllers\LoginController::loginView');
$routes->post('login', 'Auth\LoginController::loginAction');
```
